ejb3/src/main/java/org/jboss/as/ejb3/tx/CMTTxInterceptor.java

```diff
@@ -244,49 +167,74 @@ public Object processInvocation(InterceptorContext invocation) throws Exception
                     throw EjbLogger.ROOT_LOGGER.unknownTxAttributeOnInvocation(attr, invocation);
             }
         } finally {
-            tm.setTransactionTimeout(oldTimeout == -1 ? 0 : oldTimeout);
+            // See also https://issues.jboss.org/browse/WFTC-44
+            tm.setTransactionTimeout(oldTimeout == ContextTransactionManager.getGlobalDefaultTransactionTimeout() ? 0 : oldTimeout);
         }
     }
 
     protected Object invokeInCallerTx(InterceptorContext invocation, Transaction tx, final EJBComponent component) throws Exception {
         try {
             return invocation.proceed();
+        } catch (Error e) {
+            setRollbackOnly(tx);
+            throw EjbLogger.ROOT_LOGGER.unexpectedErrorRolledBack(e);
+        } catch (Exception e) {
+            ApplicationExceptionDetails ae = component.getApplicationException(e.getClass(), invocation.getMethod());
+
+            if (ae != null) {
+                if (ae.isRollback()) setRollbackOnly(tx);
+                throw e;
+            }
+            try {
+                throw e;
+            } catch (EJBTransactionRolledbackException | NoSuchEJBException | NoSuchEntityException e2) {
+                setRollbackOnly(tx);
+                throw e2;
+            } catch (RuntimeException e2) {
+                setRollbackOnly(tx);
+                throw new EJBTransactionRolledbackException(e2.getMessage(), e2);
+            }
         } catch (Throwable t) {
-            handleInCallerTx(invocation, t, tx, component);
+            throw new UndeclaredThrowableException(t);
         }
-        throw new RuntimeException("UNREACHABLE");
     }
 
     protected Object invokeInNoTx(InterceptorContext invocation, final EJBComponent component) throws Exception {
         try {
             return invocation.proceed();
+        } catch (Error e) {
+            throw EjbLogger.ROOT_LOGGER.unexpectedError(e);
+        } catch (EJBException e) {
+            throw e;
+        } catch (RuntimeException e) {
+            ApplicationExceptionDetails ae = component.getApplicationException(e.getClass(), invocation.getMethod());
+            throw ae != null ? e : new EJBException(e);
+        } catch (Exception e) {
+            throw e;
         } catch (Throwable t) {
-            handleExceptionInNoTx(invocation, t, component);
+            throw new UndeclaredThrowableException(t);
```

Comment:
This part I'm still not 100% sure about.  I could just drop the ``Throwable`` branch and everything still compiles cleanly.  And since UTE is a RuntimeException, it technically would usually qualify as a system exception and as such, it should be wrapped in EJBException.  So I could drop it, keep it as is, or wrap in ``EJBException``.

Reply:
I would probably wrap it in EJBException, but I don't think it really matters that much either way